### ISSUE TITLE
Problem with scheduler thread pools from ThreadPoolManager

### ISSUE THREAD

**Issue by bobadair** [openhab/openhab-core#942]
Hi. I’ve seen a problem in the Lutron binding under 2.5.0 snapshot builds that I’ve traced back to a problem with scheduler thread pools from ThreadPoolManager. A routine which should be periodically executed by a thread scheduled using scheduleWithFixedDelay() is never being called. It looks like when threads in the thread pool time out from being idle, new ones aren’t being started for some reason, at least under certain circumstances.

**Original issue**

The Lutron binding’s IPBridgeHandler uses the ScheduledExecutorService “scheduler” from the BaseThingHandler class for task scheduling. I believe this should end up using the “thingHandler” thread pool. When IPBridgeHandler starts up, it starts a long-running messageSender task using schedule(), and a periodic heartbeat task called keepAlive using scheduleWithFixedDelay(). There can be some other tasks scheduled in certain circumstances, but for most of the time in normal operation, it will be using just one active thread (messageSender) and have one periodic task queued (keepAlive).

If I set keepAlive to execute every 30 or 60 seconds, the thread pool will stay at a size of 3 or 4, with one thread active and one task queued, and keepAlive will always be executed as expected. However, if I set it to run at intervals of 66 seconds or more, the pool size will drop to 1 before the first keepAlive execution because of the thread idle timeout (which is hard-coded at 65 seconds in ThreadPoolManager), and the keepAlive task will never be executed. It seems like once the thread pool size shrinks, it can’t grow again for some reason, and since one thread is already active, another task won’t be run.

**Investigation**

To investigate this further I created a test binding that reproduces the problem in a much simpler context.  All it does is start a long running task with scheduler.schedule(), and then schedule a periodic task with scheduler.scheduleWithFixedDelay().  The binding has one thing *threadtest* with one parameter *interval* which sets the time interval for the periodic task in seconds. It also has one channel *poolStats* which triggers a dump of task info and scheduler pool stats when any string is sent to it.  Since the channel is defined as a `<command>` channel, this is easily done by clicking on its button under Control in Paper UI.

If *interval* is set to 66 or more seconds, the periodic task will never be executed.  The thread pool size drops to 1, and that one thread is presumably tied up executing the long running task.  More threads are never added, and the scheduled periodic task is never executed.

If *interval* is less than 65 seconds, idle threads will not time out, and the problem will not occur. Interestingly, it seems that if there is no long running task to take up a thread the problem will also not occur.

The code for the test binding is available on github [here](https://github.com/bobadair/openhab2-addons/tree/threadtestnew/bundles/org.openhab.binding.threadtestnew)

A jar file containing the test binding is available [here](https://github.com/bobadair/openhab2-addons/releases/tag/threadtest-release1)

Here is sample log output:

    10:07:02.049 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:60 - Scheduling periodic job with interval 90
    10:07:02.050 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:94 - Periodic job status: delay 89 done false cancelled false
    10:07:02.050 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:96 - Scheduler status: org.eclipse.smarthome.core.internal.common.WrappedScheduledExecutorService@53c8289d[Running, pool size = 3, active threads = 1, queued tasks = 1, completed tasks = 1] shut: false term: false
    10:07:25.586 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:94 - Periodic job status: delay 66 done false cancelled false
    10:07:25.587 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:96 - Scheduler status: org.eclipse.smarthome.core.internal.common.WrappedScheduledExecutorService@53c8289d[Running, pool size = 3, active threads = 1, queued tasks = 1, completed tasks = 1] shut: false term: false
    10:07:40.615 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:94 - Periodic job status: delay 51 done false cancelled false
    10:07:40.615 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:96 - Scheduler status: org.eclipse.smarthome.core.internal.common.WrappedScheduledExecutorService@53c8289d[Running, pool size = 3, active threads = 1, queued tasks = 1, completed tasks = 1] shut: false term: false
    10:07:50.839 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:94 - Periodic job status: delay 41 done false cancelled false
    10:07:50.839 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:96 - Scheduler status: org.eclipse.smarthome.core.internal.common.WrappedScheduledExecutorService@53c8289d[Running, pool size = 3, active threads = 1, queued tasks = 1, completed tasks = 1] shut: false term: false
    10:08:03.557 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:94 - Periodic job status: delay 28 done false cancelled false
    10:08:03.557 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:96 - Scheduler status: org.eclipse.smarthome.core.internal.common.WrappedScheduledExecutorService@53c8289d[Running, pool size = 3, active threads = 1, queued tasks = 1, completed tasks = 1] shut: false term: false
    10:08:17.951 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:94 - Periodic job status: delay 14 done false cancelled false
    10:08:17.951 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:96 - Scheduler status: org.eclipse.smarthome.core.internal.common.WrappedScheduledExecutorService@53c8289d[Running, pool size = 1, active threads = 1, queued tasks = 1, completed tasks = 1] shut: false term: false
    10:08:27.246 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:94 - Periodic job status: delay 4 done false cancelled false
    10:08:27.246 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:96 - Scheduler status: org.eclipse.smarthome.core.internal.common.WrappedScheduledExecutorService@53c8289d[Running, pool size = 1, active threads = 1, queued tasks = 1, completed tasks = 1] shut: false term: false
    10:08:34.571 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:94 - Periodic job status: delay -2 done false cancelled false
    10:08:34.571 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:96 - Scheduler status: org.eclipse.smarthome.core.internal.common.WrappedScheduledExecutorService@53c8289d[Running, pool size = 1, active threads = 1, queued tasks = 1, completed tasks = 1] shut: false term: false
    10:08:43.462 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:94 - Periodic job status: delay -11 done false cancelled false
    10:08:43.462 [safeCall-2] INFO  o.o.b.t.i.ThreadTestNewHandler:96 - Scheduler status: org.eclipse.smarthome.core.internal.common.WrappedScheduledExecutorService@53c8289d[Running, pool size = 1, active threads = 1, queued tasks = 1, completed tasks = 1] shut: false term: false

**Comment by lolodomo** [comment]
In parallel to this issue, I think an issue is to use by a binding of the common thread pool for a long-running thread. It will lock one thread of this pool. This looks like a conception issue in the binding. A solution for such binding could be to define a dedicated thread pool. If I correctly remember, some bindings are sometimes defining a dedicated thread pool.
I will have a look to the lutron binding but more generally a check of all bindings should be started...

**Comment by bobadair** [comment]
I've already changed the Lutron binding to start a Thread for the message sender rather than using schedule(). However, I believe there are other bindings that use the scheduler thread pool for long-running tasks as well.

**Comment by lolodomo** [comment]
Do you know which ones ?

**Comment by J-N-K** [comment]
OneWire probably. Will Check that.

**Comment by bobadair** [comment]
I did a quick bit of searching, and it looks like these five bindings may be starting long-running tasks using the scheduler thread pool from BaseThingHandler. I'm not familiar with them all, though, so the most I can really say is that is seems likely from a quick look.

org.openhab.binding.elerotransmitterstick/src/main/java/org/openhab/binding/elerotransmitterstick/internal/stick/TransmitterStick.java (Line 55)
```
    public synchronized void initialize(EleroTransmitterStickConfig stickConfig, ScheduledExecutorService scheduler) {
        logger.debug("Initializing Transmitter Stick...");
        config = stickConfig;
        worker = new CommandWorker();
        scheduler.schedule(worker, 0, TimeUnit.MILLISECONDS);
        logger.debug("Transmitter Stick initialized, worker running.");
    }
```

--------------

org.openhab.binding.innogysmarthome/src/main/java/org/openhab/binding/innogysmarthome/internal/handler/InnogyBridgeHandler.java (Line 659)
```
    public void onEventRunnerStopped(long delay) {
        logger.trace("onEventRunnerStopped called");
        scheduler.schedule(new WebSocketRunner(this), delay, TimeUnit.SECONDS);
    }
```

--------------

org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/MieleBridgeHandler.java (Line 552)
```
        logger.debug("Scheduling the Miele event listener job");

        if (eventListenerJob == null || eventListenerJob.isCancelled()) {
            eventListenerJob = scheduler.schedule(eventListenerRunnable, 0, TimeUnit.SECONDS);
        }
```

--------------

org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/internal/connector/telnet/DenonMarantzTelnetConnector.java (Line 71)
```
    public void connect() {
        telnetClient = new DenonMarantzTelnetClient(config, this);
        telnetRunnable = scheduler.submit(telnetClient);
    }
```

--------------

org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/transceiver/EnOceanTransceiver.java (Line 159)
```
    public void StartReceiving(ScheduledExecutorService scheduler) {

        if (readingTask == null || readingTask.isCancelled()) {
            readingTask = scheduler.submit(new Runnable() {

                @Override
                public void run() {
                    receivePackets();
                }

            });
        }
    }
```

**Comment by lolodomo** [comment]
I just discovered that the hue binding is one of the bindings using the thinghandler thread pool for a long running task:
https://github.com/openhab/openhab-addons/blob/2.5.x/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/HttpClient.java#L97

**Comment by J-N-K** [comment]
This needs to addressed in the add-on.
